sifive-blocks/templates/metal/private/metal_private_sifive_hca0.h.j2

```diff
@@ -0,0 +1,245 @@
+    uint32_t w;                    /**< Structure used for word access */
+} HCA_DMA_CR_Type;
+
+/* HCA DMA control */
+#define HCA_DMA_CR_START_Pos              0U                                       /**< HCA DMA_CR: Start Position */
+#define HCA_DMA_CR_START_Msk              (1U << HCA_DMA_CR_START_Pos)             /**< HCA DMA_CR: Start Mask */
+
+#define HCA_DMA_CR_BUSY_Pos               8U                                       /**< HCA DMA_CR: Busy Position */
+#define HCA_DMA_CR_BUSY_Msk               (1U << HCA_DMA_CR_BUSY_Pos)              /**< HCA DMA_CR: Busy Mask */
+
+#define HCA_DMA_CR_RDALIGNERR_Pos         9U                                       /**< HCA DMA_CR: Read alignment error Position */
+#define HCA_DMA_CR_RDALIGNERR_Msk         (1U << HCA_DMA_CR_RDALIGNERR_Pos)        /**< HCA DMA_CR: Read alignment error Mask */
+
+#define HCA_DMA_CR_WRALIGNERR_Pos         10U                                      /**< HCA DMA_CR: Write alignment error Position */
+#define HCA_DMA_CR_WRALIGNERR_Msk         (1U << HCA_DMA_CR_WRALIGNERR_Pos)        /**< HCA DMA_CR: Write alignment error Mask */
+
+#define HCA_DMA_CR_RESPERR_Pos            11U                                      /**< HCA DMA_CR: Response error Position */
+#define HCA_DMA_CR_RESPERR_Msk            (1U << HCA_DMA_CR_RESPERR_Pos)           /**< HCA DMA_CR: Response error Mask */
+
+#define HCA_DMA_CR_LEGALERR_Pos           12U                                      /**< HCA DMA_CR: Legal error Position */
+#define HCA_DMA_CR_LEGALERR_Msk           (1U << HCA_DMA_CR_LEGALERR_Pos)          /**< HCA DMA_CR: Legal error Mask */
+
+/**
+ * Structure type to access DMA length (DMA_LEN)
+ */
+typedef union _HCA_DMA_LEN {
+    struct {
+        uint32_t LEN:28;           /**< bit:  0..27  DMA Length (number of 128-bits blocks) */
+        uint32_t _reserved0:4;     /**< bit: 28..31  (reserved) */
+    } b;                           /**< Structure used for bit access */
+    uint32_t w;                    /**< Structure used for word access */
+} HCA_DMA_LEN_Type;
+
+/* HCA DMA length */
+#define HCA_DMA_LEN_LEN_Pos               0U                                       /**< HCA DMA_LEN: Length (number of 128-bits blocks) Position */
+#define HCA_DMA_LEN_LEN_Msk               0xFFFFFFFU                               /**< HCA DMA_LEN: Length (number of 128-bits blocks) Mask */
+
+/* HCA DMA source address */
+#define HCA_DMA_SRC_SRC_Pos               0U                                       /**< HCA DMA_SRC: Source Address Position */
+#define HCA_DMA_SRC_SRC_Msk               0xFFFFFFFFU                              /**< HCA DMA_SRC: Source Address Mask */
+
+/* HCA DMA destination address */
+#define HCA_DMA_DEST_DEST_Pos             0U                                       /**< HCA DMA_DEST: Destination Address Position */
+#define HCA_DMA_DEST_DEST_Msk             0xFFFFFFFFU                              /**< HCA DMA_DEST: Destination Address Mask */
+
```

Comment:
All of the structures/types/defines above this point look like they might be better placed directly in the driver itself or in a driver-internal header file. Unless we're expecting to need to dynamically generate any of the above based on Devicetree content, I'd rather keep just the Devicetree-derived content in templates.

This same comment applies to metal_private_sifive_hca0_aes.h.j2 and the like, which don't appear to have any Devicetree-derived content in them.

Reply:
register information from the HCA have been into sifive-block/metal/drivers directory